freedomofpress/securedrop
# App-tests in CI

## Status

:+1: Ready for review 


## Description of Changes

* Fixes #1729 - `app-test` role failing on second run of playbook against a staging server that has been rebooted.
* Addresses #1724 - Adds application tests into staging CI. Im not :100: % certain how these failing tests will affect CI passing/failing status yet. 
* Added a `tor` and `test` docker instance for utilizing tor in CI to perform checks outside of the `app` and `mon` boxes. In the future, we should plan to throw all tor specific tests that need to be run outside of the SD  CI servers on this `test` instance. The `tor` docker container is merely a socks proxy for other docker containers.
* Fixes a BIG regression with CSS files not getting propagated to the app deb package. Added accompanying testinfra tests to test for this scenario to prevent future problems.

## Testing

Well this is all specifically targeted at CI right now - I would recommend testing locally to confirm that the docker changes work against your machine (specifically you wacky Mac users :man_dancing: ). To recreate locally:

- Source the CI environment variables (make sure you change the vpc-id accordingly - youll get a prompt... youll see what i mean)
```
. ./devops/scripts/local-setup.sh
```
- Ensure your docker environment is ready to go. You GRSEC peeps -- `docker-machine` is your friend.
- Run `make ci-debug` - to prevent CI from tearing down on error
- Run `make ci-go`
- Wait until you see the process chugging along on ansible on the `built` stage (probably like 2 minutes in). Open another window, and jump into the `test` docker container and make sure you can run curl commands via the tor proxy:
```bash
$ docker exec -it test bash
cab015fde045$  curl -s --socks5-hostname \"${TOR_PROXY}\":9050 https://check.torproject.org/ | grep "Congratulations"
```
- If the above command worked tor is working from within the containers, now just wait til the rest of the tests run. XML file should dump into `junit/junit.xml` and you should see about 100 tests with the class prefix `apptest`

:tada: 

## Deployment

Any special considerations for deployment? Consider both:

N/A

## Checklist

### If you made changes to the app code:

N/A

### If you made changes to the system configuration:

- [x] Testinfra tests pass on the app-staging CI
- [x] Testinfra tests pass on the mon-staging CI
